lib/Sema/PlaygroundTransform.cpp

```diff
@@ -786,6 +799,14 @@ class Instrumenter : InstrumenterBase {
     return std::make_pair(PBD, VD);
   }
 
+  bool shouldLogType(Type Ty) {
```

Comment:
Maybe generalize this over `ASTNode` instead i.e. `shouldLog(ASTNode)` so that you can cast to `ValueDecl *` internally to get its interface type? 

```swift
bool shouldLog(ASTNode node) {
  if (auto *VD = node.dyn_cast<ValueDecl *>())
    return VD->hasInterfaceType() ? VD->getInterfaceType()->isNoncopyable() : true;
  if (auto *E = node.dyn_cast<Expr *>())
    return E->getType()->isNoncopyable();
  return true;
}
```

Reply:
I tried this change and the CI is failing with:
```
/Users/ec2-user/jenkins/workspace/swift-PR-macos-smoke-test/branch-main/llvm-project/llvm/include/llvm/ADT/STLExtras.h:172:42: error: implicit instantiation of undefined template 'llvm::FirstIndexOfType<swift::ValueDecl *>'
    : std::integral_constant<size_t, 1 + FirstIndexOfType<T, Us...>::value> {};
                                         ^
/Users/ec2-user/jenkins/workspace/swift-PR-macos-smoke-test/branch-main/llvm-project/llvm/include/llvm/ADT/STLExtras.h:172:42: note: in instantiation of template class 'llvm::FirstIndexOfType<swift::ValueDecl *, swift::CaseLabelItem *>' requested here
/Users/ec2-user/jenkins/workspace/swift-PR-macos-smoke-test/branch-main/llvm-project/llvm/include/llvm/ADT/STLExtras.h:172:42: note: in instantiation of template class 'llvm::FirstIndexOfType<swift::ValueDecl *, swift::StmtConditionElement *, swift::CaseLabelItem *>' requested here
/Users/ec2-user/jenkins/workspace/swift-PR-macos-smoke-test/branch-main/llvm-project/llvm/include/llvm/ADT/STLExtras.h:172:42: note: in instantiation of template class 'llvm::FirstIndexOfType<swift::ValueDecl *, swift::TypeRepr *, swift::StmtConditionElement *, swift::CaseLabelItem *>' requested here
/Users/ec2-user/jenkins/workspace/swift-PR-macos-smoke-test/branch-main/llvm-project/llvm/include/llvm/ADT/STLExtras.h:172:42: note: in instantiation of template class 'llvm::FirstIndexOfType<swift::ValueDecl *, swift::Pattern *, swift::TypeRepr *, swift::StmtConditionElement *, swift::CaseLabelItem *>' requested here
/Users/ec2-user/jenkins/workspace/swift-PR-macos-smoke-test/branch-main/llvm-project/llvm/include/llvm/ADT/STLExtras.h:172:42: note: in instantiation of template class 'llvm::FirstIndexOfType<swift::ValueDecl *, swift::Decl *, swift::Pattern *, swift::TypeRepr *, swift::StmtConditionElement *, swift::CaseLabelItem *>' requested here
/Users/ec2-user/jenkins/workspace/swift-PR-macos-smoke-test/branch-main/llvm-project/llvm/include/llvm/ADT/STLExtras.h:172:42: note: (skipping 1 context in backtrace; use -ftemplate-backtrace-limit=0 to see all)
/Users/ec2-user/jenkins/workspace/swift-PR-macos-smoke-test/branch-main/llvm-project/llvm/include/llvm/ADT/PointerUnion.h:230:30: note: in instantiation of template class 'llvm::FirstIndexOfType<swift::ValueDecl *, swift::Expr *, swift::Stmt *, swift::Decl *, swift::Pattern *, swift::TypeRepr *, swift::StmtConditionElement *, swift::CaseLabelItem *>' requested here
    return F.Val.getInt() == FirstIndexOfType<To, PTs...>::value;
                             ^
/Users/ec2-user/jenkins/workspace/swift-PR-macos-smoke-test/branch-main/llvm-project/llvm/include/llvm/ADT/PointerUnion.h:248:27: note: in instantiation of function template specialization 'llvm::CastInfoPointerUnionImpl<swift::Expr *, swift::Stmt *, swift::Decl *, swift::Pattern *, swift::TypeRepr *, swift::StmtConditionElement *, swift::CaseLabelItem *>::isPossible<swift::ValueDecl *>' requested here
    return Impl::template isPossible<To>(f);
                          ^
/Users/ec2-user/jenkins/workspace/swift-PR-macos-smoke-test/branch-main/llvm-project/llvm/include/llvm/Support/Casting.h:311:19: note: in instantiation of member function 'llvm::CastInfo<swift::ValueDecl *, llvm::PointerUnion<swift::Expr *, swift::Stmt *, swift::Decl *, swift::Pattern *, swift::TypeRepr *, swift::StmtConditionElement *, swift::CaseLabelItem *>>::isPossible' requested here
    if (!Derived::isPossible(f))
                  ^
/Users/ec2-user/jenkins/workspace/swift-PR-macos-smoke-test/branch-main/llvm-project/llvm/include/llvm/Support/Casting.h:406:23: note: in instantiation of member function 'llvm::DefaultDoCastIfPossible<swift::ValueDecl *, llvm::PointerUnion<swift::Expr *, swift::Stmt *, swift::Decl *, swift::Pattern *, swift::TypeRepr *, swift::StmtConditionElement *, swift::CaseLabelItem *>, llvm::CastInfo<swift::ValueDecl *, llvm::PointerUnion<swift::Expr *, swift::Stmt *, swift::Decl *, swift::Pattern *, swift::TypeRepr *, swift::StmtConditionElement *, swift::CaseLabelItem *>>>::doCastIfPossible' requested here
    return ForwardTo::doCastIfPossible(const_cast<NonConstFrom>(f));
                      ^
/Users/ec2-user/jenkins/workspace/swift-PR-macos-smoke-test/branch-main/swift/lib/Sema/PlaygroundTransform.cpp:803:25: note: in instantiation of function template specialization 'llvm::PointerUnion<swift::Expr *, swift::Stmt *, swift::Decl *, swift::Pattern *, swift::TypeRepr *, swift::StmtConditionElement *, swift::CaseLabelItem *>::dyn_cast<swift::ValueDecl *>' requested here
    if (auto *VD = node.dyn_cast<ValueDecl *>())
```

Reply:
ah, it's union with a `Decl *` so you need to do `dyn_cast_or_null<ValueDecl>(node.dyn_cast<Decl *>())`

Reply:
Thanks Pavel, CI is green with this change now. Does the rest of the code look reasonable to you?